enterprise/coderd/usage/publisher_test.go

```diff
@@ -0,0 +1,702 @@
+package usage_test
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/goleak"
+	"go.uber.org/mock/gomock"
+
+	"cdr.dev/slog/sloggers/slogtest"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbmock"
+	"github.com/coder/coder/v2/coderd/database/dbtestutil"
+	"github.com/coder/coder/v2/coderd/database/dbtime"
+	agplusage "github.com/coder/coder/v2/coderd/usage"
+	"github.com/coder/coder/v2/enterprise/coderd/coderdenttest"
+	"github.com/coder/coder/v2/enterprise/coderd/usage"
+	"github.com/coder/coder/v2/testutil"
+	"github.com/coder/quartz"
+)
+
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m, testutil.GoleakOptions...)
+}
+
+// TestIntegration tests the collector and publisher by running them with a real
+// database.
+func TestIntegration(t *testing.T) {
+	t.Parallel()
+	const eventCount = 3
+
+	ctx := testutil.Context(t, testutil.WaitLong)
+	log := slogtest.Make(t, nil)
+	db, _ := dbtestutil.NewDB(t)
+	clock := quartz.NewMock(t)
+	deploymentID, licenseJWT := configureDeployment(ctx, t, db)
+	now := time.Now()
+
+	var (
+		calls   int64
+		handler func(req usage.TallymanIngestRequestV1) any
+	)
+	ingestURL := fakeServer(t, tallymanHandler(t, licenseJWT, func(req usage.TallymanIngestRequestV1) any {
+		callCount := atomic.AddInt64(&calls, 1)
```

Comment:
In order to prevent test races, we need to ensure that the publish calls are completed before we attempt to read this value. If the test prevents such races, `atomic` is superfluous because we will not concurrently access it. Making it atomic doesn't do anything to prevent racy tests and is confusing because it signals to people that there might be concurrent accesses, when we _need to ensure there are none_.